File: src/ConsiliumTempus.Api/Contracts/ProjectTask/UpdateCustomField/UpdateCustomFieldFromProjectTaskRequest.cs
using ConsiliumTempus.Domain.Common.Enums;

namespace ConsiliumTempus.Api.Contracts.ProjectTask.UpdateCustomField;

public sealed record UpdateCustomFieldFromProjectTaskRequest(
    Guid Id,
    Guid CustomFieldId,
    CustomFieldType Type,
    UpdateCustomFieldFromProjectTaskRequest.DateCustomFieldRequest? DateCustomField,
    UpdateCustomFieldFromProjectTaskRequest.DateTimeCustomFieldRequest? DateTimeCustomField,
    UpdateCustomFieldFromProjectTaskRequest.DurationCustomFieldRequest? DurationCustomField,
    UpdateCustomFieldFromProjectTaskRequest.MultiSelectCustomFieldRequest? MultiSelectCustomField,
    UpdateCustomFieldFromProjectTaskRequest.NumberCustomFieldRequest? NumberCustomField,
    UpdateCustomFieldFromProjectTaskRequest.PeopleCustomFieldRequest? PeopleCustomField,
    UpdateCustomFieldFromProjectTaskRequest.SingleSelectCustomFieldRequest? SingleSelectCustomField,
    UpdateCustomFieldFromProjectTaskRequest.TextCustomFieldRequest? TextCustomField,
    UpdateCustomFieldFromProjectTaskRequest.TimeCustomFieldRequest? TimeCustomField)
{
    public sealed record DateCustomFieldRequest(
        DateOnly? Date);

    public sealed record DateTimeCustomFieldRequest(
        DateTime? DateTime);

    public sealed record DurationCustomFieldRequest(
        TimeSpan? Duration);

    public sealed record MultiSelectCustomFieldRequest(
        Guid OptionId,
        bool Remove);

    public sealed record NumberCustomFieldRequest(
        decimal? Number);

    public sealed record PeopleCustomFieldRequest(
        Guid? PersonId);

    public sealed record SingleSelectCustomFieldRequest(
        Guid? OptionId);

    public sealed record TextCustomFieldRequest(
        string? Text);

    public sealed record TimeCustomFieldRequest(
        TimeOnly? Time);
}

File: tests/e2e/types/requests/project-task/UpdateCustomFieldFromProjectTaskRequest.ts
export default interface UpdateCustomFieldFromProjectTaskRequest {
  id: string,
  customFieldId: string,
  type: string,
  dateCustomField?: DateCustomField,
  dateTimeCustomField?: DateTimeCustomField,
  durationCustomField?: DurationCustomField,
  multiSelectCustomField?: MultiSelectCustomField,
  numberCustomField?: NumberCustomField,
  peopleCustomField?: PeopleCustomField,
  singleSelectCustomField?: SingleSelectCustomField,
  textCustomField?: TextCustomField,
  timeCustomField?: TimeCustomField,
}

interface DateCustomField {
  date?: string
}

interface DateTimeCustomField {
  dateTime?: string
}

interface DurationCustomField {
  duration?: string
}

interface MultiSelectCustomField {
  optionId: string,
  remove: boolean
}

interface NumberCustomField {
  number?: number
}

interface PeopleCustomField {
  personId?: string
}

interface SingleSelectCustomField {
  optionId?: string,
}

interface TextCustomField {
  text?: string
}

interface TimeCustomField {
  time?: string
}